c++/kernel_overhead_cl/Mandelbrot.cpp

```diff
@@ -0,0 +1,187 @@
+#include <iostream>
+#include <iomanip>
+#include <sstream>
+#include <vector>
+#include <cassert>
+#include <boost/format.hpp>
+
+#ifdef _WIN32
+#include <sys/timeb.h>
+#include <windows.h>
+#else
+#include <sys/time.h>
+#include <unistd.h>
+#endif
+
+#include "TGA.h"
+#include "oclutils.h"
+
+inline double getCurrentTime() {
+#ifdef WIN32
+    LARGE_INTEGER f;
+    LARGE_INTEGER t;
+    QueryPerformanceFrequency(&f);
+    QueryPerformanceCounter(&t);
+    return t.QuadPart / (double)f.QuadPart;
+#else
+    struct timeval tv;
+    struct timezone tz;
+    gettimeofday(&tv, &tz);
+    return tv.tv_sec + tv.tv_usec*1e-6;
+#endif
+};
+
+std::vector<std::vector<float> > mandelbrot(
+    unsigned int nx, unsigned int ny, unsigned int iterations,
+    std::vector<float> x0, std::vector<float> y0,
+    std::vector<float> dx, std::vector<float> dy,
+    size_t block_width = 8, size_t block_height = 8) {
+    cl_int error = CL_SUCCESS;
+    int num_zooms = x0.size();
+    assert(num_zooms == x0.size());
+    assert(num_zooms == y0.size());
+    assert(num_zooms == dx.size());
+    assert(num_zooms == dy.size());
+
+    //Allocate GPU data 
+    std::vector<cl::Buffer> output_gpu(num_zooms);
+    for (int i = 0; i < num_zooms; ++i) {
+        output_gpu[i] = cl::Buffer(*OpenCLUtils::getContext(), CL_MEM_READ_WRITE,
+            nx*ny*sizeof(float), NULL, &error);
+        CL_CHECK(error);
+    }
+
+    //Create timing events
+    std::vector<cl::Event> events(num_zooms);
+
+    //Run kernel and generate images
+    cl::Kernel *kernel = OpenCLUtils::getKernel("mandelbrotKernel");
+    double enqueue_compute_start = getCurrentTime();
+
+    for (int i = 0; i < num_zooms; ++i) {
+        kernel->setArg<cl::Buffer>(0, output_gpu[i]);
+        kernel->setArg<unsigned int>(1, (unsigned int) (nx*sizeof(float)));
+        kernel->setArg<unsigned int>(2, nx);
+        kernel->setArg<unsigned int>(3, ny);
+        kernel->setArg<unsigned int>(4, iterations);
+        kernel->setArg<float>(5, x0[i]);
+        kernel->setArg<float>(6, y0[i]);
+        kernel->setArg<float>(7, dx[i]);
+        kernel->setArg<float>(8, dy[i]);
+
+        // execute kernel
+        CL_CHECK(OpenCLUtils::getQueue()->enqueueNDRangeKernel(
+                 *kernel, cl::NullRange, 
+                 cl::NDRange(((nx + block_width - 1)/block_width)*block_width, ((ny + block_height - 1)/block_height)*block_height), 
+                 cl::NDRange(block_width, block_height), 0, &events[i]));
+    }
+    double enqueue_compute_end = getCurrentTime();
+
+    //Synchronize
+    double sync_compute_start = getCurrentTime();
+    double gpu_time_compute = 0.0;
+    for (int i = 0; i < num_zooms; ++i) {
+        CL_CHECK(events[i].wait());
```

Comment:
Here we should wait for the kernel event (returned from enqueueNDRangeKernel(....) ) and check the 
CL_PROFILING_COMMAND_START and
CL_PROFILING_COMMAND_END. See the Jupyter notebook for details